tests/src/rules/no-unused-modules.typescript.js

```diff
@@ -0,0 +1,42 @@
+import { test, testFilePath } from '../utils'
+import typescriptConfig from '../../../config/typescript'
+
+import { RuleTester } from 'eslint'
+
+// We need to clear the cache to reinitialize the rule for Typescript only
```

Comment:
seems like maybe instead of requiring this (since the rule will run on both kinds of files in the same run), we should move the listFilesToProcess code in the rule from module level down to the rule logic itself?

Reply:
That would require moving more functions in the module to the logic itself. The `FileEnumerator` class [uses its own cache in the module level](https://github.com/eslint/eslint/blob/master/lib/cli-engine/file-enumerator.js#L195-L212) and that could cause an unexpected increase in memory usage. I might dig deeper to find a better solution. However, that might end in greater refactoring of the rule itself. Do you want it? Or is it a good idea for another PR instead?

Reply:
fair, let's work on that in a followup PR instead of this one.

Reply:
I have added JSX only test cases. However, I have had to do a few adjustments as well to make it work as you suggested. So consider the work half-done. I believe there is still space for improvement.